### ISSUE TITLE
Crafty RCE Designation

### ISSUE THREAD

**Issue by xithical** [chebuya/minegrief#1]
Hi, 

I'm one of the maintainers of the Crafty project and was taking a look at this due to some community reports we had concerning this. 

First, I would like to note that the module for Crafty uses old default credentials - this was replaced with randomly-generated admin credentials in the [4.2.2 release](https://gitlab.com/crafty-controller/crafty-4/-/releases/v4.2.2) due to people not following best practice and exposing instances to the Internet with default creds still in place. 

Second, I'm concerned about the designation of an "undisclosed, unpatched" RCE in Crafty. The code used appears to utilize the API endpoint we have that creates a Minecraft server. The entire functional purpose of Crafty is to build/create and maintain Minecraft servers through an easily-accessed web interface. Access to this API (and the web UI counterpart) is controlled via our RBAC system, which restricts server creation only to roles that explicitly allow it. To me at least, this is not undisclosed, nor is it unpatched. This is a core feature of the product functioning as intended and has controls in place to prevent abuse. I'm curious why this repository describes it as an "undisclosed, unpatched" RCE?

Happy to chat further on this and look forward to your response.

**Comment by chebuya** [comment]
Yes, this is why I described it as "undisclosed, unpatched, intentional(?)", as I assumed the application was developed with this intent. I also stated it requires authentication.

However, it is curious to me that the application appears to have protections against other classes of vulnerabilities such as path traversal. I have seen other applications that provide the ability to run system commands in a safer way.

Irregardless if this feature was intentional or not, it is by all definitions, a vulnerability (even if the Crafty team considers it an acceptable risk due to compensating controls).  So I stand by my definition.

**Comment by xithical** [comment]
Crafty is designed in such a way that it will only run with a non-administrative service account specifically for the purpose of preventing execution of commands as root and the execution command can only be changed by an administrative user in the panel. This is a documented and intentional feature of the product.

Pterodactyl, for example, which is an alternative product to Crafty, looks to run commands in much the same way we do; using a set execution string as the command to run a server that can only be modified by an administrative user.

I do see your reasoning, but I disagree with this being a vulnerability. By the logic given, SSH, Cockpit, and every RMM in the world is an authenticated RCE; not *wrong* per se, but is a little broad for the context of how the term vulnerability is practically used. It's not really a vulnerability in the product and is more a vulnerability in the user's implementation, i.e. if they have it exposed to the internet or use weak credentials, which can be said for just about any system that allows authed command execution.

I'm curious as to what you would consider a safer way to run the commands needed for server execution, as this is something we've been discussing internally for quite some time but have had little success with.

**Comment by chebuya** [comment]
Hello I would argue that the point of an RMM app such as cockpit is to execute commands, while the point of crafty is to manage minecraft servers and stuff like running commands would be ideally handled by the application

In terms of safer way to run commands, you need to allow the user to change arguments of the java command rather then the whole command

This is the current implementation
```python
                self.process = subprocess.Popen(
                    self.server_command,
                    cwd=self.server_path,
                    stdin=subprocess.PIPE,
                    stdout=subprocess.PIPE,
                    stderr=subprocess.STDOUT,
                    env=my_env,
                )
```

You should do something like this (https://stackoverflow.com/questions/89228/how-do-i-execute-a-program-or-call-a-system-command)
```python
subprocess.run(["java", arg1, arg2, arg3]) 
```

This will allow the panel user to add options to the java command but not run a entirely different command such as `curl`

This is argument injection vs command injection
https://cheatsheetseries.owasp.org/cheatsheets/OS_Command_Injection_Defense_Cheat_Sheet.html
https://cheatsheetseries.owasp.org/cheatsheets/OS_Command_Injection_Defense_Cheat_Sheet.html#argument-injection

**Comment by xithical** [comment]
The point of Crafty, at its core, is to launch server processes using defined execution commands, thus the analogy to products like Cockpit. Long-term, the goal with Crafty is not to just manage Minecraft servers; it is to have Crafty be a platform for managing game servers of many types. Even now, we don't just run Java Minecraft servers; currently, we also provide functionality for running Bedrock servers (which run as their own executable within the server folder) and have SteamCMD support in progress.

Bedrock in and of itself breaks the model you describe - by running a Bedrock server, we're having to run the executable from within the server directory, and that executable could theoretically be arbitrarily modified by someone with files access.

I don't really see a good way for us to add support for additional server types without building something that either is so inconvenient to the user that it restricts usability of the product or is so close to just running a defined execution command that it nullifies any benefit added by migrating away from the current solution. 

Additionally, it's not like any user is able to just modify the execution command. Modifying this requires a user account with the super user role, so effectively a privileged account.

I think rather than trying to chase the rabbit hole of restricting the execution command, which requires a privileged account and can be relatively easily circumvented anyways, we can try exploring the following avenues:
1. Enforce 2FA on all privileged accounts
2. Look into implementing a fail2ban listener/module that ties into our auth log to better defend against brute forcing
3. Implement IP allow/block lists to restrict access for both privileged and unprivileged accounts to trusted locations (maybe a separate list for each?)

Thoughts?

**Comment by imscruz** [comment]
> Crafty'nin amacı, özünde, tanımlanmış yürütme komutlarını kullanarak sunucu işlemlerini başlatmaktır, bu nedenle Cockpit gibi ürünlere benzetmedir. Uzun vadede, Crafty'nin amacı sadece Minecraft sunucularını yönetmek değil; Crafty'nin birçok türdeki oyun sunucularını yönetmek için bir platform olmasını sağlamaktır. Şimdi bile sadece Java Minecraft sunucularını çalıştırmıyoruz; şu anda, Bedrock sunucularını (sunucu klasörü içinde kendi yürütülebilir dosyaları olarak çalışan) çalıştırmak için işlevsellik de sağlıyoruz ve devam eden SteamCMD desteğine sahibiz.
> 
> Bedrock kendi içinde tanımladığınız modeli bozar - bir Bedrock sunucusu çalıştırarak, yürütülebilir dosyayı sunucu dizininin içinden çalıştırmak zorunda kalırız ve bu yürütülebilir dosya teorik olarak dosya erişimi olan biri tarafından keyfi olarak değiştirilebilir.
> 
> Kullanıcı için ürünün kullanılabilirliğini kısıtlayacak kadar rahatsız edici olan veya yalnızca tanımlanmış bir yürütme komutunu çalıştırmaya o kadar yakın olan bir şey oluşturmadan ek sunucu türleri için destek eklemenin gerçekten iyi bir yolunu görmüyorum.
> 
> Ek olarak, herhangi bir kullanıcı yalnızca yürütme komutunu değiştirebilir gibi değildir. Bunu değiştirmek için süper kullanıcı rolüne sahip bir kullanıcı hesabı, yani etkili bir şekilde ayrıcalıklı bir hesap gerekir.
> 
> Ayrıcalıklı bir hesap gerektiren ve yine de nispeten kolay bir şekilde atlatılabilen yürütme komutunu kısıtlamanın tavşan deliğini kovalamaya çalışmak yerine, aşağıdaki yolları keşfetmeyi deneyebiliriz:
> 
> 1. Tüm ayrıcalıklı hesaplarda 2FA'yı zorunlu kılın
> 2. Kaba zorlamaya karşı daha iyi savunma sağlamak için kimlik doğrulama günlüğümüze bağlanan bir fail2ban dinleyicisi/modülü uygulamaya bakın
> 3. Hem ayrıcalıklı hem de ayrıcalıksız hesapların erişimini güvenilen konumlarla kısıtlamak için IP izin verme/engelleme listeleri uygulayın (belki her biri için ayrı bir liste?)
> 
> Düşünce?

transfer domain to Cloudflare and add a chapcha
